### PR TITLE
Trace to events

### LINKED ISSUE #49: Event formation MVP

### PR DESCRIPTION
**Pull request by Modularius**
Closes #49

### CODE REVIEW

**Comment by DanNixon** [comment]
Is there no way this PR can be split up? 7k lines is way too big for a single PR.

**Comment by Modularius** [comment]
Yes, I can submit the trace-to-pulses crate separately.

**Comment by DanNixon** [comment]
A few things that would form a starting point of making this more reviewable:
- Move the code from the `trace-to-pulses` crate into `trace-to-events` (the former is just support code for the latter, so I see no real reason for the extra crate)
- Change the visibility modifiers from `pub` to `pub (crate)` (this will show you where you have code that is not actually used)
- The unused code can then be removed (this is not to say that it is not valuable, just that it does not yet belong on `main`)

**Comment by DanNixon** [comment]
> Yes, I can submit the trace-to-pulses crate separately.

That will not really help, that crate is the vast majority of the code in this PR.

**Review by DanNixon — CHANGES_REQUESTED** [review]
An initial look over. There are more places with [style](https://github.com/STFC-ICD-Research-and-Design/supermusr-data-pipeline/blob/main/docs/style.md) issues that I did not comment on.

File: trace-to-events/src/main.rs [hunk 1]
```diff
@@ -27,20 +32,23 @@ struct Cli {
...
     event_topic: String,
 
-    #[clap(long)]
-    threshold: Intensity,
-
     #[clap(long, default_value = "127.0.0.1:9090")]
     observability_address: SocketAddr,
+
+    #[clap(long)]
+    save_file_name: Option<String>,
```
[the remark this answers is not in the document]
Comment (Modularius) [reply trace-to-events/src/main.rs:48]: It is useful to have the option at this point. As flat-buffer event messages are limited in what they can capture at this point. Perhaps this should be made a dev-only feature?

File: trace-to-events/src/parameters.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+pub(crate) struct ConstantPhaseDiscriminatorParameters {
+    #[clap(
+        long,
+        help = "constant phase threshold for detecting muon events, use format (threshold,duration,cool_down). See README.md."
+    )]
+    pub(crate) threshold_trigger: ThresholdDurationWrapper,
+}
+
+pub(crate) struct SaveOptions<'a> {
+    pub(crate) save_path: &'a str,
```
[the remark this answers is not in the document]
Comment (Modularius) [reply trace-to-events/src/parameters.rs:38]: This is to ensure the save folder is created if it doesn't exist. Also it is easier to manipulate the file name if it separate. There is often more than one file created, e.g. filename_raw.csv, filename_smoothed.csv, filename_pulses.csv, for each stage of the process. As the saving feature is likely to be redundant when we've debugged it all, I don't think this is too much a problem. I've changed the Strings to PathBufs however.

[the remark this answers is not in the document]
Comment (DanNixon) [reply trace-to-events/src/parameters.rs:38]: I don't think creating directories should be the job of this tool, this goes against behaviour of a lot of other command line tools that manipulate/output files.

Perhaps it is easier to just specify an output directory.

File: trace-to-events/src/pulse_detection/detectors/basic_muon_detector.rs [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+                }
+                .map(|end| {
+                    let mut steepest_rise = steepest_rise.unwrap_or_default();
+                    let mut sharpest_fall = source
+                        .get_data()
+                        .get_superlative()
+                        //.map(|tv| tv)
+                        .unwrap_or_default();
+
+                    //if end.time - start.time <= 0 { Problem }
```
Comment (DanNixon) [inline trace-to-events/src/pulse_detection/detectors/basic_muon_detector.rs:274]: Should something be done here?

File: trace-to-events/src/pulse_detection/detectors/basic_muon_detector.rs [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+                        })
+                    }
+                    _ => None,
+                }
+                .map(|end| {
+                    let mut steepest_rise = steepest_rise.unwrap_or_default();
+                    let mut sharpest_fall = source
+                        .get_data()
+                        .get_superlative()
+                        //.map(|tv| tv)
```
Comment (DanNixon) [inline trace-to-events/src/pulse_detection/detectors/basic_muon_detector.rs:271]:
```suggestion
```

**Review by DanNixon — CHANGES_REQUESTED** [review]

**Review by DanNixon — APPROVED** [review]

**Pull request merged**